app/src/org/odk/collect/android/activities/FormEntryActivity.java

```diff
@@ -614,34 +526,233 @@ protected void onActivityResult(int requestCode, int resultCode, Intent intent)
+         * fixed, (read, we move on from Android 1.6) we want to handle images the audio and
+         * video
+         */
+
+        // get gp of chosen file
+        Uri selectedImage = intent.getData();
+        File originalImage = new File(FileUtils.getPath(this, selectedImage));
+
+        if (originalImage.exists()) {
+            File unscaledFinalImage = moveAndScaleImage(originalImage, true);
+
+            // Add the new image to the Media content provider so that the viewing is fast in Android 2.0+
+            ContentValues values = new ContentValues(6);
+            values.put(Images.Media.TITLE, unscaledFinalImage.getName());
+            values.put(Images.Media.DISPLAY_NAME, unscaledFinalImage.getName());
+            values.put(Images.Media.DATE_TAKEN, System.currentTimeMillis());
+            values.put(Images.Media.MIME_TYPE, "image/jpeg");
+            values.put(Images.Media.DATA, unscaledFinalImage.getAbsolutePath());
+
+            Uri imageURI =
+                    getContentResolver().insert(Images.Media.EXTERNAL_CONTENT_URI, values);
+            Log.i(TAG, "Inserting image returned uri = " + imageURI.toString());
+
+            ((ODKView) mCurrentView).setBinaryData(imageURI);
+            saveAnswersForCurrentScreen(DO_NOT_EVALUATE_CONSTRAINTS);
+        } else {
+            // The user has managed to select a file from the image browser that doesn't actually
+            // exist on the file system anymore
+            CommCareActivity.createErrorDialog(this, Localization.get("invalid.image.selection"), false);
+        }
+
+        refreshCurrentView();
+    }
+
+    private void processChooserResponse(Intent intent) {
+        // For audio/video capture/chooser, we get the URI from the content provider
+        // then the widget copies the file and makes a new entry in the content provider.
+        Uri media = intent.getData();
+        String binaryPath = UriToFilePath.getPathFromUri(CommCareApplication._(), media);
+        if (!FormUploadUtil.isSupportedMultimediaFile(binaryPath)) {
+            // don't let the user select a file that won't be included in the
+            // upload to the server
+            ((ODKView) mCurrentView).clearAnswer();
+            Toast.makeText(FormEntryActivity.this,
+                    Localization.get("form.attachment.invalid"),
+                    Toast.LENGTH_LONG).show();
+        } else {
+            ((ODKView) mCurrentView).setBinaryData(media);
+        }
+        saveAnswersForCurrentScreen(DO_NOT_EVALUATE_CONSTRAINTS);
+        refreshCurrentView();
+    }
+
+    // In order for this method to successfully return the "current" widget, the subclass of
+    // QuestionWidget that you wish to use it for must take in a PendingCalloutInterface in its
+    // constructor (passed from WidgetFactory) and call setPendingCalloutFormIndex() when it
+    // launches its intent (See IntentWidget or ImageWidget for examples)
+    private QuestionWidget getPendingWidget() {
```

Comment:
I actually talked with him in the process of doing this part because I saw that he had just changed it. This method is the result of combining his refactor (which introduced the PendingCalloutInterface) with what I needed, and is really the same as what he had originally besides making it return a generic QuestionWidget, so it can be used in multiple places instead of just for an IntentWidget.